---
title: Splicing events classification
author: Jean Monlong
date: February 8, 2016
output:
   md_document:
     fig_width: 10
     fig_height: 8
     variant: markdown_github
---

# Splicing events classification

## Introduction

To classify splicing events...

## Building the transcript structure

The *data.frame* with the transcript structure must have these columns:

+ `transId`: the transcript ID.
+ `strand`: the DNA strand.
+ `cdsStart`: the start positions of the CDS regions, separated by `,`.
+ `cdsEnd`: same with the end positions.
+ `urtStarts` and `utrEnds`: same with UTR regions.

Depending on the annotation you are using, you might need to build this *data.frame* manually.

Let's create a fake one, to use for this tutorial.

```{r, message=FALSE, warning=FALSE}
tr.str = data.frame(transId=c("t1","t2","t3"),
                    strand="+",
                    cdsStarts=c("10,40,100","10,20,100","10,40,100"),
                    cdsEnds=c("15,55,130","15,30,130","15,55,130"),
                    utrStarts=c("5,130","5,130","5,130"),
                    utrEnds=c("10,135","10,135","10,150"))
tr.str
```				


## Simple classification

Now we will compare the structure of transcripts pairs *t1* vs *t2*, and *t1* vs *t3*. We load the package and run:

```{r, message=FALSE, warning=FALSE}
library(sQTLseekeR)
tr.df = data.frame(transId1=c("t1","t1"), transId2=c("t2","t3"))
classify.events(tr.df, tr.str)
```

The result is a list with the event code and name for each pair, as well as a *data.frame* with the global count of each event in the data.

It found *mutually exclusive exons* between transcript *t1* and *t2*. Transcripts *t1* and *t3* have the same CDS but differ in their last UTR, hence the event found being *tandem 3' UTR*.


## Across multiple transcript pairs

In practice the input is a larger *data.frame* with information about each sQTL, including the relevant columns *transId1* and *transId2*. These two columns contains the IDs of the two transcripts whose relative expression change the most.

In the next example, we import the transcript structure of some genes, select 2 random transcript in each and compute the corresponding splicing event.

First we need to load a transcript annotation and build the transcript structure *data.frame*. In practice, we used the GTF of the version used for the project (mapping, transcript quantification). For this tutorial, we use *AnnotationHub* package to quickly retrieve an annotation and prepare the data.

```{r, message=FALSE, warning=FALSE, cache=TRUE}
library(AnnotationHub)
ah = AnnotationHub()
genc.gr = ah[["AH49010"]]
genc.gr = subset(genc.gr, gene_type=="protein_coding" & source=="ENSEMBL")
concat <- function(x) paste(x, collapse=",")
library(dplyr)
cds.gr = subset(genc.gr, type=="CDS")
cds.str = data.frame(transId=cds.gr$transcript_id, st=start(cds.gr), end=end(cds.gr), strand=strand(cds.gr)) %>%
  group_by(transId, strand) %>% summarize(cdsStarts=concat(st), cdsEnds=concat(end))
utr.gr = subset(genc.gr, type=="UTR")
utr.str = data.frame(transId=utr.gr$transcript_id, st=start(utr.gr), end=end(utr.gr), strand=strand(utr.gr)) %>%
  group_by(transId, strand) %>% summarize(utrStarts=concat(st), utrEnds=concat(end))
tr.str = merge(cds.str, utr.str)
```

Now let's pick randomly a thousand genes, and for each 2 transcipts. We also remove genes with only one transcript.

```{r, message=FALSE, warning=FALSE}
gene.tr = unique(mcols(subset(genc.gr, type=="transcript"))[,c("gene_id","transcript_id")])
gene.tr = as.data.frame(subset(gene.tr, gene_id %in% sample(unique(gene_id),1000)))
tr.df = gene.tr %>% group_by(gene_id) %>% mutate(transcript_id=sample(transcript_id)) %>%
  summarize(transId1=transcript_id[1], transId2=transcript_id[2]) %>% filter(transId1!=transId2)
```

At this point we are ready, we have our two *data.frames*:

```{r, message=FALSE, warning=FALSE}
str(tr.df)
str(tr.str)
```

We run the function the same way as before.

```{r, message=FALSE, warning=FALSE}
ev.l = classify.events(tr.df, tr.str)
head(ev.l$res)
head(ev.l$stats)
```

Eventually we can plot the distribution of the events in this data.

```{r, message=FALSE, warning=FALSE}
library(ggplot2)
ggplot(ev.l$stats, aes(x=event, y=prop.sqtl)) + geom_bar(stat="identity") + coord_flip() + ylab("proportion of transcript pairs")
```
